crates/sozo/src/commands/dev.rs

```diff
@@ -0,0 +1,265 @@
+                        }
+                    }
+                }
+            }
+            DevAction::None
+        }
+        _ => DevAction::None,
+    };
+    action
+}
+
+struct DevContext<'a> {
+    pub db: RootDatabase,
+    pub unit: CompilationUnit,
+    pub ws: Workspace<'a>,
+}
+
+fn load_context(config: &Config) -> Result<DevContext> {
+    let ws = scarb::ops::read_workspace(config.manifest_path(), config)?;
+    let packages: Vec<scarb::core::PackageId> = ws.members().map(|p| p.id).collect();
+    let resolve = scarb::ops::resolve_workspace(&ws)?;
+    let compilation_units = scarb::ops::generate_compilation_units(&resolve, &ws)?
+        .into_iter()
+        .filter(|cu| packages.contains(&cu.main_package_id))
+        .collect::<Vec<_>>();
+    // we have only 1 unit in projects
+    let unit = compilation_units.get(0).unwrap();
+    let db = build_scarb_root_database(&unit, &ws).unwrap();
+    Ok(DevContext { db, unit: unit.clone(), ws })
+}
+
+fn build(context: &mut DevContext) -> Result<()> {
+    let ws = &context.ws;
+    let unit = &context.unit;
+    let package_name = unit.main_package_id.name.clone();
+    ws.config().compilers().compile(unit.clone(), &mut ((*context).db), ws).map_err(|err| {
+        ws.config().ui().anyhow(&err);
+
+        anyhow!("could not compile `{package_name}` due to previous error")
+    })?;
+    ws.config().ui().print("📦 Rebuild done");
+    Ok(())
+}
+
+async fn migrate<P, S>(
+    mut world_address: Option<FieldElement>,
+    account: &SingleOwnerAccount<P, S>,
+    name: Option<String>,
+    ws: &Workspace<'_>,
+    previous_manifest: Option<Manifest>,
+) -> Result<(Manifest, Option<FieldElement>)>
+where
+    P: Provider + Sync + Send + 'static,
+    S: Signer + Sync + Send + 'static,
+{
+    let target_dir = ws.target_dir().path_existent().unwrap();
+    let target_dir = target_dir.join(ws.config().profile().as_str());
+    let manifest_path = target_dir.join("manifest.json");
+    if !manifest_path.exists() {
+        return Err(anyhow!("manifest.json not found"));
+    }
+    let new_manifest = Manifest::load_from_path(manifest_path)?;
+    let diff = WorldDiff::compute(new_manifest.clone(), previous_manifest);
+    let total_diffs = diff.count_diffs();
+    let config = ws.config();
+    config.ui().print(format!("Total diffs found: {total_diffs}"));
+    if total_diffs == 0 {
+        return Ok((new_manifest, world_address));
+    }
+    match migration::apply_diff(target_dir, diff, name.clone(), world_address, account, config)
+        .await
+    {
+        Ok(address) => {
+            config.ui().print(format!("🎉 World at address {} updated!", format!("{:#x}", address)));
+            world_address = Some(address);
+        }
+        Err(err) => {
+            config.ui().error(err.to_string());
+            return Err(err);
+        }
+    }
+
+    Ok((new_manifest, world_address))
+}
+
+impl DevArgs {
+    pub fn run(self, config: &Config) -> Result<()> {
```

Comment:
can we refactor this to improve readability?

maybe something like:

```
fn process_event(event: &DebouncedEvent, context: &mut Context) -> DevAction {
    let action = handle_event(event);
    match &action {
        DevAction::None => {}
        DevAction::Build(path) => handle_build_action(path, context),
        DevAction::Reload => handle_reload_action(context),
    }
    action
}

fn handle_build_action(path: &Path, context: &mut Context) {
    context.ws.config().ui().print(format!(
        "📦 Need to rebuild {}",
        path.to_str().unwrap_or_default(),
    ));
    let db = &mut context.db;
    let file = FileId::new(db, path);
    PrivRawFileContentQuery::in_db_mut(db.as_files_group_mut()).invalidate(&file);
    db.override_file_content(file, None);
}

fn handle_reload_action(context: &mut Context) {
    context.ws.config().ui().print("Reloading project");
    context = load_context(config).expect("Failed to load context");
    env_metadata = dojo_metadata_from_workspace(&context.ws).and_then(|inner| inner.env().cloned());
}

// Inside the loop:
loop {
    let action = match rx.recv() {
        Ok(Ok(events)) => events.iter().map(|event| process_event(event, &mut context)).last().unwrap_or(DevAction::None),
        Ok(Err(_)) => DevAction::None,
        Err(error) => {
            log::error!("Error: {:?}", error);
            break;
        }
    };

    if action != DevAction::None {
        if let Ok(_) = build(&mut context) {
            if let Err(error) = context.ws.config().tokio_handle().block_on(migrate(
                world_address,
                &account,
                name.clone(),
                &context.ws,
                previous_manifest.clone(),
            )) {
                log::error!("Error: {:?}", error);
            }
        }
    }
}
```

Reply:
@tarrencev I agree but I didn't yet succeed this refactoring due to mutable borrow issues.

I hope I could make it tomorrow.